core/shared/src/main/scala/scalaz/zio/IO.scala

```diff
@@ -173,16 +163,42 @@ sealed abstract class IO[+E, +A] extends Serializable { self =>
    * then the action will fail with some error.
    */
   final def raceBoth[E1 >: E, B](that: IO[E1, B]): IO[E1, Either[A, B]] =
-    raceWith(that)((a, fiber) => fiber.interrupt.const(Left(a)), (b, fiber) => fiber.interrupt.const(Right(b)))
+    raceWith(that)(_.join.map(Left(_)) <* _.interrupt, _.join.map(Right(_)) <* _.interrupt)
 
   /**
    * Races this action with the specified action, invoking the
    * specified finisher as soon as one value or the other has been computed.
    */
-  final def raceWith[E1 >: E, A1 >: A, B, C](
+  final def raceWith[E1, E2, B, C](
     that: IO[E1, B]
-  )(finishLeft: (A1, Fiber[E1, B]) => IO[E1, C], finishRight: (B, Fiber[E1, A1]) => IO[E1, C]): IO[E1, C] =
-    new IO.Race[E1, A1, B, C](self, that, finishLeft, finishRight)
+  )(
+    leftWins: (Fiber[E, A], Fiber[E1, B]) => IO[E2, C],
+    rightWins: (Fiber[E1, B], Fiber[E, A]) => IO[E2, C]
+  ): IO[E2, C] = {
+    def arbiter[E0, E1, A, B](
+      f: (Fiber[E0, A], Fiber[E1, B]) => IO[E2, C],
+      winner: Fiber[E0, A],
+      loser: Fiber[E1, B],
+      race: RefM[IO.Race],
+      done: Promise[E2, C]
+    )(res: ExitResult[E0, A]): IO[Nothing, _] =
+      race
+        .modify(
+          r =>
+            IO.when(r.won(res.succeeded))(f(winner, loser).run.flatMap(done.done(_)).void)
+              .map(_ -> r.next(res.succeeded))
+        )
+
+    (for {
+      done  <- Promise.make[E2, C]
+      race  <- RefM[IO.Race](IO.Race.Started)
+      left  <- self.fork
+      right <- that.fork
+      _     <- left.observe.flatMap(arbiter(leftWins, left, right, race, done)).fork
+      _     <- right.observe.flatMap(arbiter(rightWins, right, left, race, done)).fork
+      c     <- done.get
+    } yield c).supervised
```

Comment:
I don't think `supervised` works if the fiber is interrupted. We should probably fix that. If you fix that, this will be equivalent to the old way which used `ensuring` to make sure that if the parent is killed, the racing fibers are killed as well. 

Also in this method you can kind of see why we wanted `observe` (well, the old version, `onComplete`) to take a call back: because we turn right around and fork it because we just want to do something when observation happens. (But nonetheless, probably better to express it this way, it's more orthogonal, and we can always optimize RTS for this pattern.)

Reply:
I'm fairly sure .supervised works on interruption, at least that's pretty much the only case I'm using it for.

Reply:
Yeah I think it works, see:
https://github.com/scalaz/scalaz-zio/blob/989ca743611c82357efcdaa3ff9249d64a01eec7/core/jvm/src/main/scala/scalaz/zio/RTS.scala#L542-L546
The supervisor is run as a finalizer.

Reply:
Seems like it would be impossible to implement [`Concurrent.racePair`](https://github.com/typelevel/cats-effect/blob/master/core/shared/src/main/scala/cats/effect/Concurrent.scala#L217) on top of `raceWith` if it's supervised, the loser fiber has to stay alive some more after it exits